docs: add auto import info about shared utils (#30858)
nuxt/nuxt: docs/2.guide/2.directory-structure/1.shared.md

Code in the `shared/` directory cannot import any Vue or Nitro code.
::

::warning

> Auto-imports are not enabled by default in Nuxt v3 to prevent breaking changes in existing projects.

Reviewer: `Auto-imports are not enabled by default in Nuxt v3` 
It's not by default. We don't enable it for V3

Reviewer 2: you can enable them in v3 by setting compatibilityVersion.

Reviewer: Oh I get it now. IMO it's a bit hard to draw a clear line between feature available in comp v4 and those in the "standard" configuration when saying that it can be enabled in V3.
We can change it later if some misunderstand this sentence.

## Usage

**Method 1:** Named export

```ts twoslash [shared/utils/capitalize.ts]
export const capitalize = (input: string) => {
  return input[0] ? input[0].toUpperCase() + input.slice(1) : ''
}
```

**Method 2:** Default export

```ts twoslash [shared/utils/capitalize.ts]
export default function (input: string) {
  return input[0] ? input[0].toUpperCase() + input.slice(1) : ''
}
```

> You can now use [auto-imported](/docs/guide/directory-structure/shared#auto-imports) utility functions in `.js`, `.ts` and `.vue` files within your Vue app and the `server/` directory.

Reviewer: ```suggestion
You can now use [auto-imported](/docs/guide/directory-structure/shared#auto-imports) utilities within your Vue app and the Nitro part of Nuxt.
```

Reviewer: `the Nitro part of Nuxt.` i'm not quite satisfied with this phrasing.... this could be improved

Author: I suggest : `utilities in your Nuxt app and server/ directory.`its more  than just Nitro

## Usage

> Files placed in `shared/utils/` or `shared/types/` are **automatically imported** and can be used in your application **without explicit imports**.

Reviewer 3: suggestion: there is a reference to the way component auto imports work further down in the file (documenting how files in nested directories are **not** auto-imported).

Could this info be surfaced here instead so the reader doesn't need to continue scanning down the page?

Author: You're right. The logic and repetition of information needs to be reviewed.

**Here are my changes :**

“Usage” section should not give details of the auto import. 
Only usage. 
_(I justadded a redirection on the word “auto-imported” so that it scrolls lower)_ :

![Capture d’écran 2025-02-05 à 20 42 35](https://github.com/user-attachments/assets/cc978333-7ba8-4475-8371-eb5ce85830cb)

Here, we have details on "Auto Imports" section, with warning about app folder :

![Capture d’écran 2025-02-05 à 20 42 43](https://github.com/user-attachments/assets/cd980e06-a321-4a98-b994-d2944e81941c)

Reviewer 3: Lgtm. For now it's "alert heavy" but once v4 is live the compat. flag verbiage can be removed.

You can now use auto-imported utility functions in `.js`, `.ts` and `.vue` files within your Vue app and the `server/` directory.

**Method 1:** Named export

```ts twoslash [shared/utils/capitalize.ts]
export const capitalize = (input: string) => {
  return input[0] ? input[0].toUpperCase() + input.slice(1) : ''
}
```

**Method 2:** Default export

```ts twoslash [shared/utils/capitalize.ts]
export default function (input: string) {
  return input[0] ? input[0].toUpperCase() + input.slice(1) : ''
}
```

::warning

> Auto-import for files in `shared/utils/` is not enabled by default in the `app/` directory in Nuxt 3. If you want to use these functions in `app/`, you must [set `compatibilityVersion: 4` in your `nuxt.config.ts`](/docs/getting-started/upgrade#opting-in-to-nuxt-4).

Reviewer 4: This refers to the `app/` directory in Nuxt 3 which doesn't have one, so this is still misleading in a way.

Author: "Auto-import for files in `shared/utils/` is not enabled by default in Nuxt 3. To use these functions in `app/`, you must first set `compatibilityVersion: 4` in your `nuxt.config.ts`.
This is part of Nuxt's progressive compatibility features in preparation for Nuxt 4." 

@IlyaSemenov it's better said like that?Cliquez pour utiliser cette solution

Reviewer 4: This still sounds a bit odd. Why does the note refer app/ at all? These functions are not in app/ but in shared/, aren't they? I'd simply write "To use these functions, you must..."

Author: Thanks for your feedback! I've updated the warning to remove the reference to the app/ directory, as you suggested. 

The new text now reads: "To use these functions, you must first set `compatibilityVersion: 4` in your `nuxt.config.ts`. This is part of Nuxt's progressive compatibility features in preparation for Nuxt 4.

Only files in the `shared/utils/` and `shared/types/` directories will be auto-imported. Files nested within subdirectories of these directories will not be auto-imported.

::warning

> To use these functions auto-imported, you must first [set `compatibilityVersion: 4` in your `nuxt.config.ts`](/docs/getting-started/upgrade#opting-in-to-nuxt-4).  

Reviewer: ```suggestion
To use auto-imported shared utils and types, you must first [set `future.compatibilityVersion: 4` in your `nuxt.config.ts`](/docs/getting-started/upgrade#opting-in-to-nuxt-4).  
```

Author: yep it's better, I just remove "shared" : `To use these auto-imported utils and types`

> This is part of Nuxt's progressive compatibility features in preparation for Nuxt 4.

Reviewer: ```suggestion
The reason this isn't enabled in V3 is to prevent breaking changes amongs existing projects.
```
